source/common/http/alternate_protocols.h

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "envoy/common/time.h"
+
+namespace Envoy {
+namespace Http {
+
+// Tracks alternate protocols that can be used to make an HTTP connection to an origin server.
+class AlternateProtocols {
+public:
+  // Represents an HTTP origin to be connected too.
+  struct Origin {
+  public:
+    Origin(std::string scheme, std::string hostname, int port);
+
+    bool operator==(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) ==
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator!=(const Origin& other) const { return !this->operator==(other); }
+
+    bool operator<(const Origin& other) const {
```

Comment:
nit: should we also do `<=`, `>`,  and `>=` for convenience?

Reply:
I can't imagine ever using them, but it's trivial to add so sure, done!